### ISSUE TITLE
Any basic example of reading basic of using memory data?

### ISSUE THREAD

**Issue by FrankEscobar** [mkazhdan/PoissonRecon#90]
I mean is there any small code snippet for using an array of vertex positions in memory instead reading from an external .ply file?

Thank you in advance.

**Comment by mkazhdan** [comment]
I just pushed a slightly updated version of the code to the repo. The new PoissonRecon and SSDRecon implementations support a "--inCore" flag that allows you to keep the points in memory instead of repeatedly streaming them in from disk. Hopefully, the new code provides what you are looking for.

**Comment by FrankEscobar** [comment]
To be honest the massive usage of templates makes this really flexible but hard to follow I was asking about having a float* that stores x,y,z per vertex or something like that.

**Comment by mkazhdan** [comment]
I see. So let's assume you have an array, in, of floats with size count*6, storing x,y,z,nx,ny,nz. Then the way you construct an input point stream would look something like:
	
{
	// Define the auxiliary data stored with each point (i.e. the normals)
	typedef MultiPointStreamData< float , PointStreamNormal< float , 3 > > NormalData;

	// Move the input data into a vector of position/normal pairs
	std::vector< std::pair< Point< float , 3 > , NormalData > > out;

	// Copy the input data
	out.resize( count );
	for( int i=0 ; i<count ; i++ )
	{
		Point< float , 3 > p( in[i*6+0] , in[i*6+1] , in[i*6+2] ) , n( in[i*6+3] , in[i*6+4] , in[i*6+5] );

		// Copy in the position
		out[i].first = p;
		// Copy in the normal
		// This is a bit hairy, but the gist is as follows:
		// The data stored in the second component of the pair is a (one-)tuple whose first entry
		// is the normal data. We extract the first entry using std::get< 0 > and then access the
		// data member.
		std::get< 0 >( out[i].second.data ).data = n;
	}

	// Create the (memory-backed) input point stream.
	InputPointStreamWithData< float , 3 , NormalData > *pointStream = new MemoryInputPointStreamWithData< float , 3 , NormalData >( out.size() , &out[0] );
}

**Comment by FrankEscobar** [comment]
Thank you since I want to exted PoissonRecon I tried to use that InputPointStreamWithData* as pointStream but it seems that I cannot cast _InputPointStreamWithData< float , 3 , NormalData > *pointStream_ to _InputPointStreamWithData*_ any tip on this?

**Comment by mkazhdan** [comment]
The problem is that InputPointStreamWithData is typedef'ed to be:
	_typedef InputPointStreamWithData< Real , Dim , TotalPointSampleData > InputPointStream;_

If you look at the implementation of main, you'll see two different calls to the Execute function, with correspondingly different types of point data.
	_if( Colors.set ) Execute< Real , PointStreamColor< DefaultFloatType > >( argc , argv , FEMSigs() );
	else             Execute< Real >( argc , argv , FEMSigs() );_
The latter is the one you want since it says that (aside from normals) there is no other auxiliary data associated with point samples. Which is precisely what _InputPointStreamWithData< float , 3 , NormalData >_ describes.
The problem is that the first line tries to create a point stream that has both normals and colors as auxiliary data, which is not the same as _InputPointStreamWithData< float , 3 , NormalData >_. I believe if you dis-allow the use of colors (i.e. just keep the "else" part) things should work.

**Comment by FrankEscobar** [comment]
Unfortunatly I think that it has nothing to do with that, I made this simple test:

`
InputPointStreamWithData< float, 3, NormalData >* createMemoryStreamFromPointer(std::vector<float> &positions, std::vector<float> &normals)
{
	// Move the input data into a vector of position/normal pairs
	std::vector< std::pair< Point< float, 3 >, NormalData > > out;

	assert(positions.size() == normals.size());

	int count = positions.size();
	// Copy the input data
	out.resize(count);
	for (int i = 0; i < count; i++)
	{
		int offset = i * 3;
		Point< float, 3 > p(positions[offset + 0], positions[offset + 1], positions[offset + 2]), n(normals[offset + 0], normals[offset + 1], normals[offset + 2]);

		// Copy in the position
		out[i].first = p;
		// Copy in the normal
		// This is a bit hairy, but the gist is as follows:
		// The data stored in the second component of the pair is a (one-)tuple whose first entry
		// is the normal data. We extract the first entry using std::get< 0 > and then access the
		// data member.
		std::get< 0 >(out[i].second.data).data = n;
	}

	positions.clear();
	normals.clear();

	// Create the (memory-backed) input point stream.
	InputPointStreamWithData< float, 3, NormalData > *pointStream = new MemoryInputPointStreamWithData< float, 3, NormalData >(out.size(), &out[0]);
	return pointStream;
}

InputPointStreamWithData<float, 3, NormalData>* createBasicExample()
{
	std::vector<float> positions;
	std::vector<float> normals;

	// V0
	positions.push_back(-10);
	positions.push_back(-5);
	positions.push_back(0);
	normals.push_back(-1.5708f);
	normals.push_back(-1.5708f);
	normals.push_back(-1.5708f);
	// V1
	positions.push_back(-10);
	positions.push_back(5);
	positions.push_back(0);
	normals.push_back(-1.5708f);
	normals.push_back(1.5708f);
	normals.push_back(-1.5708f);
	// V2
	positions.push_back(10);
	positions.push_back(5);
	positions.push_back(0);
	normals.push_back(1.5708f);
	normals.push_back(1.5708f);
	normals.push_back(-1.5708f);
	// V3
	positions.push_back(10);
	positions.push_back(-5);
	positions.push_back(0);
	normals.push_back(1.5708f);
	normals.push_back(-1.5708f);
	normals.push_back(-1.5708f);
	// V4
	positions.push_back(-10);
	positions.push_back(-5);
	positions.push_back(30);
	normals.push_back(-1.5708f);
	normals.push_back(-1.5708f);
	normals.push_back(1.5708f);
	// V5
	positions.push_back(-10);
	positions.push_back(5);
	positions.push_back(30);
	normals.push_back(-1.5708f);
	normals.push_back(1.5708f);
	normals.push_back(1.5708f);
	// V6
	positions.push_back(10);
	positions.push_back(5);
	positions.push_back(30);
	normals.push_back(1.5708f);
	normals.push_back(1.5708f);
	normals.push_back(1.5708f);
	// V7
	positions.push_back(10);
	positions.push_back(-5);
	positions.push_back(30);
	normals.push_back(1.5708f);
	normals.push_back(-1.5708f);
	normals.push_back(1.5708f);

	return createMemoryStreamFromPointer(positions, normals);
}
`

And in PoissonRecon.cpp I removed 
`*if( InCore.set )
		{
			InputPointStream *_pointStream;
...`

And I replacet it with 
`
		pointStream = createBasicExample();`

And the error is 

> cannot convert from 'InputPointStreamWithData<float,3,NormalData> *' to 'InputPointStream *'

**Comment by mkazhdan** [comment]
It still depends on the template parameters with which execute is called from the "main" function. If the code supporting color is still active, the conversion should fail.

**Comment by FrankEscobar** [comment]
I have no more references to color in my code also I removed them from PointStreamData.h but without luck

**Comment by FrankEscobar** [comment]
Hello again @mkazhdan I hope you had a nice Christmas!

I'll try to continue my static implementation of your project but I'm still having issues

After removing any reference to color I still have this issue:
`error C2440: '=': cannot convert from 'InputPointStreamWithData<float,3,NormalData> *' to 'InputPointStream *'`

on 

`InputPointStream* pointStream = createBasicExample();`

Any tip on this?

**Comment by mkazhdan** [comment]
Hi Frank,

I think at this point it might be easier if we could look at your code. Is that something you can share? (If so, you can email it to me directly misha@cs.jhu.edu.)

-- Misha

**Comment by ashf** [comment]
I'm looking at implementing something similar.  I'd like to pass data from my program directly instead of via a PLY file. You already helped me with getting the reconstruction library to function as a re-entrant DLL. This is the next frontier.

I'm trying to follow the example. But I'm having some trouble. Unlike the OP, I do want color information, and I can't seem to figure out how to set that into this line:
`InputPointStreamWithData< float, 3, NormalData > *pointStream = new MemoryInputPointStreamWithData< float, 3, NormalData >(out.size(), &out[0]);`
NOTE: I actually changed that line to:
`InputPointStreamWithData< float, 3, Point< float, 3> > *pointStream = new MemoryInputPointStreamWithData< float, 3, Point< float, 3> >(out.size(), &out[0]);`

**Comment by BlaseL** [comment]
Admittedly, the code is a bit messy here.
Basically, when you construct an InputPointStreamWithData object, you need to specify:
-- Real: the type floating point you're using
-- Dim: the dimension of the point, and 
-- TotalPointSampleData: the data (aside from the point position) that is packaged with the point.
Typically, TotalPointSampleData is an object of type MultiPointStreamData which essentially a bit tuple.
For the specific case of Poisson reconstruction we require normals, we make TotalPointSampleData be a 2-tuple:
-- NormalPointSampleData: The first element is the normal
-- AdditionalPointSampleData: The second is itself MultiPointStreamData object (i.e. a tuple).
To get things to work, you need to "package" the color information within the inessential data, AdditionalPointSampleData.

**Comment by mkazhdan** [comment]
Admittedly, the code is a bit convoluted here.
Basically, when you construct an InputPointStreamWithData object, you need to specify:
-- Real: the type floating point you're using
-- Dim: the dimension of the point, and
-- TotalPointSampleData: the data (aside from the point position) that is packaged with the point.
Typically, TotalPointSampleData is an object of type MultiPointStreamData which is essentially a tuple.
For the specific case of Poisson reconstruction we require normals, so we make TotalPointSampleData be a 2-tuple consisting of two elements:
-- NormalPointSampleData: The normal information
-- AdditionalPointSampleData: A MultiPointStreamData object (i.e. a tuple) storing all the other information.
(So you get a 2-tuple on of whose elements is a tuple.)
To get things to work, you need to "package" the color information within the non-normal data wrapper, AdditionalPointSampleData.

**Comment by ashf** [comment]
I meant to write, that I did get it working. Tuple madness, I think in the end I had to do a .data.data.data.data :)

**Comment by mkazhdan** [comment]
Glad you got it working.
Your're that the tuple madness is kind of ugly. (I've been working on a slightly cleaner interface which halves the number of "data" terms.)

**Comment by DamonsJ** [comment]
std::vector< std::pair< Point< Real, Dim >, TotalPointSampleData > > inCorePoints;
        if (InCore.set)
        {
            std::vector<Real> points_normals = { 0,0,0,0,0,1,1,0,0,0,0,1 ,1,1,0,0,0,1 ,0,1,0,0,0,1 };
            int count = points_normals.size()/6;
            // Copy the input data
            inCorePoints.resize(count);
            for (int i = 0; i < count; i++)
            {
                Point< float, 3 > p(points_normals[i * 6 + 0], points_normals[i * 6 + 1], points_normals[i * 6 + 2]), n(points_normals[i * 6 + 3], points_normals[i * 6 + 4], points_normals[i * 6 + 5]);
                // Copy in the position
                inCorePoints[i].first = p;
                // Copy in the normal
                // This is a bit hairy, but the gist is as follows:
                // The data stored in the second component of the pair is a (one-)tuple whose first entry
                // is the normal data. We extract the first entry using std::get< 0 > and then access the
                // data member.
                std::get< 0 >(inCorePoints[i].second.data()).data() = n;
            }

            pointStream = new MemoryInputPointStreamWithData< Real, Dim, TotalPointSampleData >(inCorePoints.size(), &inCorePoints[0]);
        }

just replace original code in PoissonRecon.cpp at line 510 
and using inCore mode

I get right result using my code

**Comment by FrankEscobar** [comment]
Here we go again updating the question according to the new implementations.

Also I think that keeping an update sample could be nice for future users too.

In this case I started coding a basic sphere point cloud:
```

void createSphereSample(std::vector<double> &pos, std::vector<double> &nor)
{
	int sphereSubdivs = 500;
	double sphereRadius = 10.0;
	const double sectorStep = 2 * 3.14159265359 / sphereSubdivs;
	double stackStep = 3.14159265359 / sphereSubdivs;
	for (int i = 1; i < sphereSubdivs; ++i)
	{
		const double Phi = i * stackStep - 0.5 * 3.14159265359;
		const double cosPhi = cos(Phi);
		const double z = sin(Phi);

		for (int j = 0; j < sphereSubdivs; ++j)
		{
			const double Theta = j * sectorStep;
			const double x = cosPhi * cos(Theta); 
			const double y = cosPhi * sin(Theta);

			pos.push_back(sphereRadius * x);
			nor.push_back(x);
			pos.push_back(sphereRadius * y);
			nor.push_back(y);
			pos.push_back(sphereRadius * z);
			nor.push_back(z);
		}
	}
	printf("pos %i (%i) nor %i (%i)\n", pos.size() / 3, pos.size(), nor.size() / 3, nor.size());
}
```

Nothing too fancy but enough for testing.


In the function _void Execute( UIntPack< FEMSigs ... > , const AuxDataFactory &auxDataFactory )_

I've replaced
`if( InCore.set )`

by
```
if (true)
{
  std::vector<double> pos;
  std::vector<double> nor;
  createSphereSample(nor, pos);
  for (int mv = 0; mv < pos.size()/3; mv++)
  {
    const int mv3 = mv*3;
    InputSampleType pointAndNormal;
    pointAndNormal.get<0>() = Point< Real, Dim >(pos[mv3], pos[mv3+1], pos[mv3+2]);
    pointAndNormal.get<1>().get<0>() = Point< Real, Dim >(nor[mv3], nor[mv3 + 1], nor[mv3 + 2]);
    inCorePoints.push_back(pointAndNormal);
  }
  pointStream = new MemoryInputDataStream< InputSampleType >(inCorePoints.size(), &inCorePoints[0]);  
  printf("here we go!\n");
}
else if( InCore.set )
```

But unfortunately it didn't work

> [ERROR] D:\Tests\PoissonReconLib\Src\Geometry.h (Line 222)
> Point<float>::operator +=
> Dimensions don't match: 5 != 0

Any tip? Thank you in advance

**Comment by jiapei-nexera** [comment]
It looks **InputPointStreamWithData** is **NOT** there any more???
